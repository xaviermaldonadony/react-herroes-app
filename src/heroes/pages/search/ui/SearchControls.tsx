import { Button } from '@/components/ui/button';
import { Input } from '@/components/ui/input';
import { Filter, Grid, Plus, Search, SortAsc } from 'lucide-react';

export const SearchControls = () => {
  return (
    <>
      <div className='flex flex-col lg:flex-row gap-4 mb-8'>
        {/* Search */}
        <div className='relative flex-1'>
          <Search className='absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-5 w-5' />
          <Input
            placeholder='Search heroes, villains, powers, teams...'
            className='pl-12 h-12 text-lg bg-white'
          />
        </div>

        {/* Action buttons */}
        <div className='flex gap-2'>
          <Button variant='outline' className='h-12 bg-transparent'>
            <Filter className='h-4 w-4 mr-2' />
            Filters
          </Button>

          <Button variant='outline' className='h-12 bg-transparent'>
            <SortAsc className='h-4 w-4 mr-2' />
            Sort by Name
          </Button>

          <Button variant='outline' className='h-12 bg-transparent'>
            <Grid className='h-4 w-4' />
          </Button>

          <Button className='h-12'>
            <Plus className='h-4 w-4 mr-2' />
            Add Character
          </Button>
        </div>
      </div>

      {/* Advanced Filters */}
      <div className='bg-white rounded-lg p-6 mb-8 shadow-sm border'>
        <div className='flex justify-between items-center mb-4'>
          <h3 className='text-lg font-semibold'>Advanced Filters</h3>
          <Button variant='ghost'>Clear All</Button>
        </div>
        <div className='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-4'>
          <div className='space-y-2'>
            <label className='text-sm font-medium'>Team</label>
            <div className='h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm'>
              All teams
            </div>
          </div>
          <div className='space-y-2'>
            <label className='text-sm font-medium'>Category</label>
            <div className='h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm'>
              All categories
            </div>
          </div>
          <div className='space-y-2'>
            <label className='text-sm font-medium'>Universe</label>
            <div className='h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm'>
              All universes
            </div>
          </div>
          <div className='space-y-2'>
            <label className='text-sm font-medium'>Status</label>
            <div className='h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm'>
              All statuses
            </div>
          </div>
        </div>
        <div className='mt-4'>
          <label className='text-sm font-medium'>Minimum Strength: 0/10</label>
          <div className='relative flex w-full touch-none select-none items-center mt-2'>
            <div className='relative h-2 w-full grow overflow-hidden rounded-full bg-secondary'>
              <div
                className='absolute h-full bg-primary'
                style={{ width: '0%' }}
              />
            </div>
            <div className='block h-5 w-5 rounded-full border-2 border-primary bg-background ring-offset-background transition-colors' />
          </div>
        </div>
      </div>
    </>
  );
};
