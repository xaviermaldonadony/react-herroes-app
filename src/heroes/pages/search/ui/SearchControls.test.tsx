import { describe, expect, test } from 'vitest';
import { SearchControls } from './SearchControls';
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

if (typeof window.ResizeObserver === 'undefined') {
  class ResizeObserver {
    observe() {}
    unobserve() {}
    disconnect() {}
  }
  window.ResizeObserver = ResizeObserver;
}

const renderWithRouter = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <SearchControls />
    </MemoryRouter>
  );
};
describe('SearchControls', () => {
  test('should render SearchControls with default values', () => {
    const { container } = renderWithRouter();

    expect(container).toMatchSnapshot();
  });

  test('should set input value when search param name is set', () => {
    renderWithRouter(['/?name=Batman']);

    const input = screen.getByPlaceholderText(
      'Search heroes, villains, powers, teams...'
    );

    expect(input.getAttribute('value')).toBe('Batman');
  });

  test('should change params when input is changed and enter is pressed', () => {
    renderWithRouter(['/?name=Batman']);
    const input = screen.getByPlaceholderText(
      'Search heroes, villains, powers, teams...'
    );

    expect(input.getAttribute('value')).toBe('Batman');

    fireEvent.change(input, { target: { value: 'Superman' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(input.getAttribute('value')).toBe('Superman');
  });

  test('should change params strength when slider is changed', () => {
    renderWithRouter(['/?name=Batman&active-accordion=advanced-filters']);
    const slider = screen.getByRole('slider');

    expect(slider.getAttribute('aria-valuenow')).toBe('0');
    fireEvent.keyDown(slider, { key: 'ArrowRight' });

    expect(slider.getAttribute('aria-valuenow')).toBe('1');
  });

  test('should accordion be open when active-accordion param is set', () => {
    renderWithRouter(['/?name=Batman&active-accordion=advanced-filters']);

    const accordion = screen.getByTestId('accordion');
    const accordionItem = accordion.querySelector('div');

    expect(accordionItem?.getAttribute('data-state')).toBe('open');
  });

  test('should accordion be open when active-accordion param is not set', () => {
    renderWithRouter(['/?name=Batman']);

    const accordion = screen.getByTestId('accordion');
    const accordionItem = accordion.querySelector('div');

    expect(accordionItem?.getAttribute('data-state')).toBe('closed');
  });
});
