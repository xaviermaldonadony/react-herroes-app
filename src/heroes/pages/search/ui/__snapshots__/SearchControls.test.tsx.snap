// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchControls > should render SearchControls with default values 1`] = `
<div>
  <div
    class="flex flex-col lg:flex-row gap-4 mb-8"
  >
    <div
      class="relative flex-1"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-search absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 h-5 w-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-4.34-4.34"
        />
        <circle
          cx="11"
          cy="11"
          r="8"
        />
      </svg>
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex w-full min-w-0 rounded-md border px-3 py-1 shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive pl-12 h-12 text-lg bg-white"
        data-slot="input"
        placeholder="Search heroes, villains, powers, teams..."
        value=""
      />
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-funnel h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20a1 1 0 0 0 .553.895l2 1A1 1 0 0 0 14 21v-7a2 2 0 0 1 .517-1.341L21.74 4.67A1 1 0 0 0 21 3H3a1 1 0 0 0-.742 1.67l7.225 7.989A2 2 0 0 1 10 14z"
          />
        </svg>
        Filters
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-arrow-up-narrow-wide h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m3 8 4-4 4 4"
          />
          <path
            d="M7 4v16"
          />
          <path
            d="M11 12h4"
          />
          <path
            d="M11 16h7"
          />
          <path
            d="M11 20h10"
          />
        </svg>
        Sort by Name
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-grid3x3 lucide-grid-3x3 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            width="18"
            x="3"
            y="3"
          />
          <path
            d="M3 9h18"
          />
          <path
            d="M3 15h18"
          />
          <path
            d="M9 3v18"
          />
          <path
            d="M15 3v18"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 px-4 py-2 has-[>svg]:px-3 h-12"
        data-slot="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus h-4 w-4 mr-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
        Add Character
      </button>
    </div>
  </div>
  <div
    data-orientation="vertical"
    data-slot="accordion"
    data-testid="accordion"
  >
    <div
      class="border-b last:border-b-0"
      data-orientation="vertical"
      data-slot="accordion-item"
      data-state="closed"
    >
      <div
        aria-labelledby="radix-«r0»"
        class="data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down overflow-hidden text-sm"
        data-orientation="vertical"
        data-slot="accordion-content"
        data-state="closed"
        hidden=""
        id="radix-«r1»"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
  </div>
</div>
`;
